1\. Write a function that converts numbers among the bin, dec, and hex representations (bin<->dec<->hex)

In [514]:
a = 5
def myConvert(a):
    a_int = int(a) #getting all types and convert to integer which is a decimal representation
    a_bin = bin(a) #similarly, converting binary representation 
    a_hex = hex(a) #similarly, converting hexadecimal representation
    print((a_int, a_bin, a_hex)) 
myConvert(a)    

(5, '0b101', '0x5')


2\. Write a function that converts a 32 bit word into a single precision floating point (i.e. interprets the various bits as sign, mantissa and exponent) and check its validity.

In [515]:
import bitstring
from decimal import Decimal


myInput = '00000011111000000000000000000000' # 32 bit word input
#print(myArr)

def myConvert2(myInput):
    e = 0
    M = 0
    myArr = bitstring.BitArray(bin = myInput).bin
    # print(myInput) 
    # Calculation of the Exponent:
    #print(myArr[0])
    for i in range(1,9):
        #print(int(i))
        e += int(myArr[i])*(2**(8-i))
    #print(e)
    # Calculation of Mantissa
    for j in range(1,23):
        #print(j)
        #print(Decimal(2**-j))
        #print((2**-j)*float(myArr[8+j]))
        M += (2**-j)*float(myArr[8+j])
    M += 1 # Out of the loop, M is increased by 1 with respect to the Mattissa Formula given.
    #print(M)    
    #print(Decimal(myArr[0]) == Decimal(0))
    #print(0)
    #print(myArr[0])
    if Decimal(myArr[0]) == Decimal(0): # Control operations by the help of Decimal. Otherwise, equality is not satisfied.
        return M*(2**(e-127))
        
    else:
        return (-M)*(2**(e-127))
    
    
    
myConvert2(myInput)


1.316553672920962e-36

3\. Write a program to determine the underflow and overflow limits (within a factor of 2) for python on your computer. 

**Tip**: define two variables initialized to 1 and halve/double them for a sufficient amount of times to exceed the under/over-flow limits  

In [516]:
import sys
import decimal
X_under = 1
def defineLimit():
    isOver = True
    isUnder = True
    X_over = 1
    X_under = 1


    while isUnder:
        myLowerLimit = X_under/2
        #print(X_under,myLowerLimit)
        if myLowerLimit == 0.0:
            print("UnderFlowLimit is reached when the number is",X_under)
            isUnder = False
        X_under/=2
        #print(X_under,myLowerLimit)

    try:
        while isOver:
            myOverLimit = X_over*2.0
            X_over*=2
            myLowerLimit = X_under/2.0
            X_under/=2
    except:
        #print("Oops!", sys.exc_info()[0], "occurred.")
        print("OverFlowLimit is reached when the number is: ",X_over)


defineLimit()

UnderFlowLimit is reached when the number is 5e-324
OverFlowLimit is reached when the number is:  179769313486231590772930519078902473361797697894230657273430081157732675805500963132708477322407536021120113879871393357658789768814416622492847430639474124377767893424865485276302219601246094119453082952085005768838150682342462881473913110540827237163350510684586298239947245938479716304835356329624224137216


4\. Write a program to determine the machine precision

**Tip**: define a new variable by adding an increasingly smaller value (in the same way as the previous problem) and check when the addition starts to have no effect on the number

In [517]:
def myPrecision():
    i=1
    count = 0
    A = 5
    A_origin = A
    A_old = A
    areTheSame = True
    while areTheSame:
        A += 0.1**i
        print(A_old,A)
        if A==A_old:
            print("A is not changing any more!!!",A_old,A)
            areTheSame=False
        A_old+=0.1**i
        #print("Just for check",A_old,A)
        i+=1
    myStr = str(A)
    print("My machine precision is: %d decimal points" %(len(myStr)-2))
        
myPrecision()        
        

5 5.1
5.1 5.109999999999999
5.109999999999999 5.111
5.111 5.1110999999999995
5.1110999999999995 5.111109999999999
5.111109999999999 5.111110999999999
5.111110999999999 5.1111111
5.1111111 5.1111111099999995
5.1111111099999995 5.111111111
5.111111111 5.1111111111
5.1111111111 5.11111111111
5.11111111111 5.111111111111
5.111111111111 5.1111111111111
5.1111111111111 5.11111111111111
5.11111111111111 5.111111111111111
5.111111111111111 5.111111111111111
A is not changing any more!!! 5.111111111111111 5.111111111111111
My machine precision is: 15 decimal points


5. Write a function that takes in input three parameters $a$, $b$ and $c$ and prints out the two solutions to the quadratic equation $ax^2+bx+c=0$ using the standard formula: $$x=\frac{-b\pm\sqrt{b^2-4ac}}{2a}$$

(a) use the program to compute the solution for $a=0.001$, $b=1000$ and $c=0.001$
(b) re-express the standard solution formula by multiplying the numerator and the denominator by $-b\mp\sqrt{b^2-4ac}$ and again find the solution for $a=0.001$, $b=1000$ and $c=0.001$. How does it compare with what has been previously obtained, and why?
(c) write a function that computes the roots of a quadratic equation accurately in all cases

In [518]:
import math
a=0.001
b=1000
c=0.001
#print(b)
#print(b**2-4*a*c)
#print(-b+math.sqrt(b**2-(4*a*c)))
#print((-b+math.sqrt(b**2-(4*a*c)))/(2*a))

def findRoots(a,b,c):
    x1 = (-b + math.sqrt((b**2)-(4*a*c)))/(2*a)
    x2 = (-b - math.sqrt((b**2)-(4*a*c)))/(2*a)
    
    return x1,x2

findRoots(a,b,c)





(-9.999894245993346e-07, -999999.999999)

In [519]:
import math

a=0.001
b=1000
c=0.001

def findRoots2(a,b,c):
    x1 = (2*c)/(-b-math.sqrt((b**2)-(4*a*c)))
    x2 = (2*c)/(-b+math.sqrt((b**2)-(4*a*c)))
    
    return x1,x2 

findRoots2(a,b,c)

(-1.000000000001e-06, -1000010.5755125057)

Modified Formula is: $$x=\frac{2c}{(-b\pm\sqrt{b^2-4ac})}$$



Since we subtracted nearly equal numbers(b and sqrt(b^2-4ac)) in the first example, it leads to loss of precision.
It gives the correct answer for the first root but still there is the same problem: "Subtraction of the nearly equal numbers."


In [520]:
import math

a=0.001
b=1000
c=0.001
Delta = b**2 - 4*a*c
#print(Delta)
def findRoots3(a,b,c):
    if b<0: # This condition always give the bigger root first!
        x1 = (2*c)/(-b+math.sqrt((b**2)-(4*a*c)))
    else:
        x1 = (2*c)/(-b-math.sqrt((b**2)-(4*a*c)))

# We know that multiplication of x1 and x2 is c/a.
        x2 = c/(a*x1)
    return x1,x2
findRoots3(a,b,c)

(-1.000000000001e-06, -999999.9999989999)

# 6\. Write a program that implements the function $f(x)=x(x−1)$

(a) Calculate the derivative of the function at the point $x = 1$ using the derivative definition:

$$
\frac{{\rm d}f}{{\rm d}x} = \lim_{\delta\to0} \frac{f(x+\delta)-f(x)}{\delta}
$$

with $\delta = 10^{−2}$. Calculate the true value of the same derivative analytically and compare it with the answer your program gives. The two will not agree perfectly. Why?

(b) Repeat the calculation for $\delta = 10^{−4}, 10^{−6}, 10^{−8}, 10^{−10}, 10^{−12}$ and $10^{−14}$. How does the accuracy scales with $\delta$?

In [521]:
# Analytically, f'(x) is equal to 1 at teh point x=1.

def myDer(a,d: float):
    # where d is very small number
    myResult = (((a+d)*(a+d-1)-a*(a-1)))/d #from the formula given.
    print(myResult)
    return myResult


myDer(1,0.01)
myDer(1,0.0001)
myDer(1,0.00001)
myDer(1,0.000001)
myDer(1,0.0000001)
myDer(1,0.00000001)
myDer(1,0.000000001)

# Derivative of f(x) is defined as how the function value f(x) changes with the infinitesimally small change of x.
# Therefore, the results are as expected because the small d values give more accurate value of f'(x)
# The concept behind is similar to previous question. When d is gradually decreased, the addition of a and d will
# be more meaningful because one of them is a high value but the other is not.
# As we know, the subtraction of nearly equal numbers leads to precision loss.

1.010000000000001
1.0000999999998899
1.0000100000065513
1.0000009999177333
1.0000001005838672
1.0000000039225287
1.000000083740371


1.000000083740371

Derivative of f(x) is defined as how the function value f(x) changes with the infinitesimally small change of x.

Therefore, the results are as expected because the small d values give more accurate value of f'(x)

The concept behind is similar to previous question. When d is gradually decreased, the addition of a and d will

be more meaningful because one of them is a high value but the other is not compared to other.

As we know, the subtraction of nearly equal numbers leads to precision loss.

7\. Consider the integral of the semicircle of radius 1:
$$
I=\int_{-1}^{1} \sqrt(1-x^2) {\rm d}x
$$
which is known to be $I=\frac{\pi}{2}=1.57079632679...$.

Alternatively we can use the Riemann definition of the integral:
$$
I=\lim_{N\to\infty} \sum_{k=1}^{N} h y_k 
$$

with $h=2/N$ the width of each of the $N$ slices the domain is divided into and where
$y_k$ is the value of the function at the $k-$th slice.

(a) Write a program to compute the integral with $N=100$. How does the result compare to the true value?

(b) How much can $N$ be increased if the computation needs to be run in less than a second? What is the gain in running it for 1 minute? 


In [512]:
import math


def mySemiCircleArea(N):
    myTotal_Y = 0
    for k in range(1,N//2+1):
        #print(k)
        myTotal_Y += math.sqrt((N-2*k)*(N+2*k))
        
    myArea = 4*myTotal_Y/(N**2)
    
    return myArea

print(mySemiCircleArea(1000000))

#%timeit mySemiCircleArea(1000000)

    

1.5707943251318113


First of all, I would like to write the formula for the quarter circle and finally multiply by 2.

h = 2/N 

For the sum, I used Nu=50 for the quarter circle. However, N will be 100 for the function definition.



$$
                I = 4\sum_{k=1}^{Nu=50}\frac{h\sqrt{1 - (2k)^2}}{N^2}  
$$

By arranging the equation h=2/N, I obtained:


$$
                I = \frac{4}{N^2}\sum_{k=1}^{Nu=50}\sqrt{(N-2k)*(N+2k)}  
$$             

In this equation N is a big integer compared to k which is also an integer. No problem with the precision.

When N is equal to a million, total time is 127ms

However, when N is equal to 10 million, total time is 1.24 seconds.

Maximum value of N should be a million for running time less than a second.

And the result is I = 1.5707943251318113 

However, when I run the timeit function, my computer temperature increased. So, I didn' t complete it.

when N is a billion, the result is correct including 8 decimal points, which is more precise as expected. When N 

increases, the error due to Riemann sum definition(the areas of the rectangles) decreases and division of the sum 

expression result in more precise number. 

$$
                I = \frac{order of (N)}{order of (N^2)}   
$$


Denominator increases more compared to nominator. Therefore, the division gives more precise number with the gradual increase of N.



In [510]:
import time
import math


def mySemiCircleArea(N):
    
    myTotal_Y = 0
    for k in range(1,N//2+1):
        #print(k)
        myTotal_Y += math.sqrt((N-2*k)*(N+2*k))
        
    myArea = 4*myTotal_Y/(N**2)
    return myArea



for i in range(100,10000000,100000):
    start_time = time.time()
    mySemiCircleArea(i)  
    finish_time = time.time()-start_time
    print(finish_time,i)
    if finish_time>=1:
        #print("--- %s seconds ---" % (finish_time))
        print("max N is: %d and the time is :%2.f " %(i,finish_time))
        print(mySemiCircleArea(i))
        break
    
    
    

1.6927719116210938e-05 100
0.015336751937866211 100100
0.02823495864868164 200100
0.04454922676086426 300100
0.05498099327087402 400100
0.09335613250732422 500100
0.08165502548217773 600100
0.10255289077758789 700100
0.10839700698852539 800100
0.12063384056091309 900100
0.12777495384216309 1000100
0.14600491523742676 1100100
0.16077804565429688 1200100
0.1831660270690918 1300100
0.1982870101928711 1400100
0.20099806785583496 1500100
0.2074289321899414 1600100
0.22803187370300293 1700100
0.2449169158935547 1800100
0.254166841506958 1900100
0.26670193672180176 2000100
0.2673637866973877 2100100
0.28145384788513184 2200100
0.30226588249206543 2300100
0.3094968795776367 2400100
0.33831238746643066 2500100
0.3503432273864746 2600100
0.369952917098999 2700100
0.3601100444793701 2800100
0.38359785079956055 2900100
0.3953237533569336 3000100
0.4091782569885254 3100100
0.4168229103088379 3200100
0.4355931282043457 3300100
0.4563479423522949 3400100
0.4593238830566406 3500100
0.47126197814941406

Time is arranged to 1 second and corresponding N is maximum 7900000. Of course more exact results can be found but it is enough to see the order of N and result of the function is approximated to analytical value.   